docs/structured-streaming-programming-guide.md

```diff
@@ -0,0 +1,888 @@
+
+Furthermore this model naturally handles data that has arrived later than expected based on its event-time. Since Spark is updating the Result Table, it has full control over updating/cleaning up the aggregates when there is late data. While not yet implemented in Spark 2.0 yet, event-time watermarking will be used to manage this data. These are explained in more details in the Window Operations section.
+
+## Fault Tolerance Semantics
+Delivering end-to-end exactly-once semantics was one of key goals behind the design of Structured Streaming. To achieve that, we have designed the Structured Streaming sources, the sinks and the execution engine to reliably track the exact progress of the processing so that it can handle any kind of failure by restarting and/or reprocessing. Every streaming source is assumed to have offsets (similar to Kafka offsets, or Kinesis sequence numbers)
+for track the read position in the stream. The engine uses checkpointing and write ahead logs to record the offset range of the data being processed in each trigger. The streaming sinks are designed to be idempotent for handling reprocessing. Together, Structured Streaming can ensure **end-to-end exactly-once semantics** under any failure.
+
+# API using Datasets and DataFrames
+Since Spark 2.0, DataFrames and can represent of static, bounded data, as well as streaming, unbounded data. Similar to static Datasets/DataFrames, you can use the common entry point  SparkSession to create straming DataFrames/Datasets from streaming sources, and apply the same operations on them as static DataFrames/Datasets. If you are not familiar with Datasets/DataFrames, you are strongly advised to familiarize yourself with them using the Spark SQL Programming Guide.
+
+## Creating streaming DataFrames and streaming Datasets
+Streaming DataFrames can created through the DataStreamReader interface `SparkSession.readStream()`. Similar to the read interface for creating static DataFrame, you can specify the details of the source - data format, schema, options, etc. In Spark 2.0, there are a few built-in sources.
+
+  - **File sources** - Reads files written in a directory as a stream of data. Supported file formats are text, csv, json, parquet. See the docs of the DataStreamReader interface for more up-to-date list, and supported options for each file format. Note that the files needs to be atomically put in the directory. In most file systems, this means that the files should be moved atomically into the directory.
```

Comment:
for more up-to-date list, => for a more up-to-date list

files needs to be atomically => files need to be atomically